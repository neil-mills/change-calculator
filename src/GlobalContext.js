import React, { createContext, useState } from 'react';
import currencies from './data/currencies';

const initialState = {
  calculateChange: () => {},
  results: null,
};

export const GlobalContext = createContext(initialState);

export const GlobalProvider = ({ children, ...rest }) => {
  const [results, setResults] = useState(initialState.results);

  const calculateChange = ({ currency, price, tendered, json }) => {
    const prefix = {
      gbp: '£',
      usd: '$',
      eur: '€',
    };

    const setPrefix = (value) => `${prefix[currency]}${value}`;

    const totalChange = parseFloat(tendered - price).toFixed(2);
    let balance = Math.round(totalChange * 100);
    const change = currencies
      .filter((c) => c.currency === currency)
      .sort((a, b) => b.denomination - a.denomination)
      .reduce((res, tender) => {
        const { title, denomination } = tender;
        const total = Math.floor(balance / denomination);
        if (total > 0) {
          balance -= total * denomination;
          res = [...res, { title, total }];
        }
        return res;
      }, []);
      const jsonChange = json ? JSON.stringify(change) : null;
    setResults({
      change,
      jsonChange,
      totalChange: setPrefix(totalChange),
      price: setPrefix(price),
      tendered: setPrefix(tendered),
    });
  };

  const values = {
    ...initialState,
    calculateChange,
    results,
    ...rest,
  };

  return (
    <GlobalContext.Provider value={values}>{children}</GlobalContext.Provider>
  );
};
