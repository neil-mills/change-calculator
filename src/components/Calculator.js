import React, { useContext } from 'react';
import { GlobalContext } from '../GlobalContext';
import { useForm } from '../hooks/useForm';
const Calculator = () => {
  const { calculateChange } = useContext(GlobalContext);

  const shape = {
    currency: {
      validate: {
        type: 'string',
        required: true,
      },
      message: 'Select a currency',
      default: '',
    },
    price: {
      validate: {
        type: 'number',
        required: true,
        lte: 'tendered'
      },
      message: 'Price is invalid',
      default: '',
    },
    tendered: {
      validate: {
        type: 'number',
        required: true,
        gte: 'price',
      },
      message: 'Amount tendered is invalid',
      default: '',
    },
    json: {
      default: false
    }
  };

  const submitForm = (values) => {
    if (!errors.length) {
      calculateChange(values);
    }
  };

  const { values, handleChange, handleSubmit, errors, isSubmitted } = useForm(
    submitForm,
    shape
  );

  const getErrorMessage = (name) =>
    errors.find((err) => err.path === name)?.message || 'Invalid';
  const isValid = (name) =>
    isSubmitted ? !errors.find((err) => err.path === name) : true;

  return (
    <div className="calculator box" data-testid="calculator">
      <form onSubmit={handleSubmit}>
        <div className="field">
          <label htmlFor="currency" className="label is-medium">
            Currency
          </label>
          <div className="control">
            <select
              name="currency"
              id="currency"
              className={`select is-medium ${
                !isValid('currency') ? 'is-danger' : ''
              }`}
              value={values.currency}
              onChange={handleChange}
              data-testid="currency-select"
            >
              <option value="">Select currency</option>
              <option value="gbp">GBP</option>
              <option value="usd">USD</option>
              <option value="eur">EUR</option>
            </select>
          </div>
          {!isValid('currency') && (
            <p className="help is-danger" data-testid="currency-error">{getErrorMessage('currency')}</p>
          )}
        </div>
        <div className="field">
          <label htmlFor="price" className="label is-medium">
            Product price
          </label>
          <div className="control">
            
            <input
              type="text"
              name="price"
              id="price"
              className={`input is-medium ${
                !isValid('price') ? 'is-danger' : ''
              }`}
              placeholder="0.00"
              value={values.price}
              onChange={handleChange}
              data-testid="price-input"
            />
          </div>
          {!isValid('price') && (
            <p className="help is-danger" data-testid="price-error">{getErrorMessage('price')}</p>
          )}
        </div>
        <div className="field">
          <label htmlFor="tendered" className="label is-medium">
            Tendered amount
          </label>
          <div className="control">
            <input
              type="text"
              name="tendered"
              id="tendered"
              className={`input is-medium ${
                !isValid('tendered') ? 'is-danger' : ''
              }`}
              placeholder="0.00"
              value={values.tendered}
              onChange={handleChange}
              data-testid="tendered-input"
            />
          </div>
          {!isValid('tendered') && (
            <p className="help is-danger" data-testid="tendered-error">{getErrorMessage('tendered')}</p>
          )}
        </div>
        <div className="field">
        <label className="checkbox" htmlFor="json">
  <input type="checkbox" name="json" onChange={handleChange} />
  Output results as JSON
</label>
        </div>
        <div className="field">
          <div className="control">
            <button type="submit" className="button is-link is-medium" data-testid="submit-button">
              Calculate
            </button>
          </div>
        </div>
      </form>
    </div>
  );
};
export default Calculator;
