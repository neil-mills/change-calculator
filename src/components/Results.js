import React, { useContext, Fragment } from 'react';
import { GlobalContext } from '../GlobalContext';

const Results = () => {
  const { results } = useContext(GlobalContext);
  if (!results) return null;
  const { change = [], totalChange = '', price = '', tendered = '', jsonChange = null } = results;
  if (jsonChange) {
    return <p className="title is-4">Coming soon...</p>
  }
  return (
    <section className="box">
      <p className="title is-4">{`You supplied ${tendered}, with a product price of ${price}`}</p>
      {change.length > 0 && (
        <Fragment>
          <p className="title is-5">This is your change:</p>
          <ul data-testid="results" className="title is-5">
            {change.map(({ title, total }, i) => (
              <li key={i} data-testid="result-item">
                <span>{`${total}x ${title}`}</span>
              </li>
            ))}
          </ul>
        </Fragment>
      )}
      <p className="title is-4">Total change: {totalChange}</p>
    </section>
  );
};

export default Results;
