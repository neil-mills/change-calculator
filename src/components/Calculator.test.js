import React from 'react';
import { render, fireEvent, cleanup, getAllByTestId } from '@testing-library/react';
import Calculator from './Calculator';
import Results from './Results';
import { GlobalProvider } from '../GlobalContext';

const renderComponent = (children, context = {}) =>
  render(
    <GlobalProvider {...context}>
      {children}
    </GlobalProvider>
  );

describe('Calculator', () => {
  beforeEach(() => {
    cleanup();
  });

  it('renders the component', () => {
    const { asFragment } = renderComponent(<Calculator />);
    expect(asFragment()).toMatchSnapshot();
  });

  describe('Validation', () => {
    const calculateChange = jest.fn();
    const mockContext = {
      calculateChange,
      results: [],
    };

    it('does not display error messages on mount', () => {
      const { queryByTestId } = renderComponent(<Calculator />, mockContext);
      expect(queryByTestId('currency-error')).toBeFalsy();
      expect(queryByTestId('price-error')).toBeFalsy();
      expect(queryByTestId('tendered-error')).toBeFalsy();
    });

    it('shows error and does not submit if no currency selected', () => {
      const { getByTestId } = renderComponent(<Calculator />,mockContext);
      fireEvent.click(getByTestId('submit-button'));
      expect(getByTestId('currency-error')).toBeTruthy();
      expect(calculateChange).toHaveBeenCalledTimes(0);
    });

    it('shows error and does not submit if price value is invalid', () => {
      const { getByTestId } = renderComponent(<Calculator />, mockContext);
      const input = getByTestId('price-input');
      const submit = getByTestId('submit-button');
      fireEvent.click(submit);
      const error = getByTestId('price-error');
      expect(error).toBeTruthy();
      expect(calculateChange).toHaveBeenCalledTimes(0);
      fireEvent.change(input, { target: { value: '2.nn' } });
      fireEvent.click(submit);
      expect(error).toBeTruthy();
      expect(calculateChange).toHaveBeenCalledTimes(0);
    });

    it('shows error and does not submit if tendered value is invalid', () => {
      const { getByTestId } = renderComponent(<Calculator />, mockContext);
      const input = getByTestId('tendered-input');
      const submit = getByTestId('submit-button');
      fireEvent.click(submit);
      const error = getByTestId('tendered-error');
      expect(error).toBeTruthy();
      expect(calculateChange).toHaveBeenCalledTimes(0);
      fireEvent.change(input, { target: { value: '2.nn' } });
      fireEvent.click(submit);
      expect(error).toBeTruthy();
      expect(calculateChange).toHaveBeenCalledTimes(0);
    });

    it('shows errors if tendered amount is less than price', () => {
      const { getByTestId } = renderComponent(<Calculator />, mockContext);
      const price = getByTestId('price-input');
      const tendered = getByTestId('tendered-input');
      fireEvent.change(price, { target: { value: '10.00' } });
      fireEvent.change(tendered, { target: { value: '5.00' } });
      fireEvent.click(getByTestId('submit-button'));
      const priceError = getByTestId('price-error');
      const tenderedError = getByTestId('tendered-error');
      expect(priceError).toBeTruthy();
      expect(tenderedError).toBeTruthy();
      expect(calculateChange).toHaveBeenCalledTimes(0);
    });

    it('submits if all form values are valid', () => {
      const { getByTestId } = renderComponent(<Calculator />, mockContext);
      const currency = getByTestId('currency-select');
      const price = getByTestId('price-input');
      const tendered = getByTestId('tendered-input');
      fireEvent.change(currency, { target: { value: 'gbp' } });
      fireEvent.change(price, { target: { value: '9.99' } });
      fireEvent.change(tendered, { target: { value: '20.00' } });
      fireEvent.click(getByTestId('submit-button'));
      expect(calculateChange).toHaveBeenCalledTimes(1);
    });
  });
  describe('Integration tests', () => {
    it('displays the correct results', () => {
      const { getByTestId, getAllByTestId, debug } = renderComponent(<div><Calculator /><Results /></div>);
      const currency = getByTestId('currency-select');
      const price = getByTestId('price-input');
      const tendered = getByTestId('tendered-input');
      fireEvent.change(currency, { target: { value: 'gbp' } });
      fireEvent.change(price, { target: { value: '11.99' } });
      fireEvent.change(tendered, { target: { value: '20.00' } });
      fireEvent.click(getByTestId('submit-button'));
      const results = getAllByTestId('result-item');
      debug();
      expect(results.length).toEqual(4);

    })
  })
});
