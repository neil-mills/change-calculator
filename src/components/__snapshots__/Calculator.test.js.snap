// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator renders the component 1`] = `
<DocumentFragment>
  <div
    class="calculator box"
    data-testid="calculator"
  >
    <form>
      <div
        class="field"
      >
        <label
          class="label is-medium"
          for="currency"
        >
          Currency
        </label>
        <div
          class="control"
        >
          <select
            class="select is-medium "
            data-testid="currency-select"
            id="currency"
            name="currency"
          >
            <option
              value=""
            >
              Select currency
            </option>
            <option
              value="gbp"
            >
              GBP
            </option>
            <option
              value="usd"
            >
              USD
            </option>
            <option
              value="eur"
            >
              EUR
            </option>
          </select>
        </div>
      </div>
      <div
        class="field"
      >
        <label
          class="label is-medium"
          for="price"
        >
          Product price
        </label>
        <div
          class="control"
        >
          <input
            class="input is-medium "
            data-testid="price-input"
            id="price"
            name="price"
            placeholder="0.00"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="field"
      >
        <label
          class="label is-medium"
          for="tendered"
        >
          Tendered amount
        </label>
        <div
          class="control"
        >
          <input
            class="input is-medium "
            data-testid="tendered-input"
            id="tendered"
            name="tendered"
            placeholder="0.00"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="field"
      >
        <label
          class="checkbox"
          for="json"
        >
          <input
            name="json"
            type="checkbox"
          />
          Output results as JSON
        </label>
      </div>
      <div
        class="field"
      >
        <div
          class="control"
        >
          <button
            class="button is-link is-medium"
            data-testid="submit-button"
            type="submit"
          >
            Calculate
          </button>
        </div>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
